stylo: Allow computing change hints during the traversal.

emilio opened the pull request:
<!-- Please describe your changes on the following line: -->

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] These changes do not require tests because geckolib :-(

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

r? @bholley 
cc @heycam

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12645)

<!-- Reviewable:end -->

emilio committed:
style: Fix sequential traversal.

emilio committed:
stylo: Allow computing change hints during the traversal.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/gecko_glue.rs, components/style/sequential.rs, components/style/dom.rs, components/style/traversal.rs, components/style/matching.rs
 * @kichjang: components/script_layout_interface/restyle_damage.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

bholley commented on components/style/gecko_glue.rs:
```diff
@@ -47,6 +46,10 @@ impl<GeckoType, ServoType> ArcHelpers<GeckoType, ServoType> {
         unsafe { transmute(owned) }
     }
 
+    pub unsafe fn borrow(borrowed: &Arc<ServoType>) -> *const &mut GeckoType {
```
Can we make this a closure-y API like |with| to avoid the unsafety here?

bholley commented:
Generally seems reasonable, but will need some fiddling in light of https://bugzilla.mozilla.org/show_bug.cgi?id=1290335#c9

emilio committed:
style: Refactor TRestyleDamage and TNode to accept/provide a "style source"

In the Gecko case, this style source would be the style context. In the servo
case, it will be always the computed values.

We could optimise this further in the case of stylo (from three FFI calls to
one) if we use an API of the form CalcAndStore(node, new_cv). But that would
imply borrowing the data twice from Servo (we also have borrow_data_unchecked
fwiw, but...).

emilio committed:
Update bindings

bholley commented on components/style/dom.rs:
```diff
@@ -46,7 +46,23 @@ impl OpaqueNode {
+    /// ComputedValues in Servo and a StyleContext in Gecko.
+    ///
+    /// This must be obtained via the node.
+    type CascadeStyleSource;
```
Let's call this PreExistingComputedValues?

bholley replied:
Also, please add a comment explaining why we do this dance with the style context (i.e. the inference from the caching).

bholley commented on components/style/dom.rs:
```diff
@@ -159,6 +175,13 @@ pub trait TNode : Sized + Copy + Clone {
+    /// XXX: It's a bit unfortunate we need to pass the current computed values
+    /// as an argument here, but otherwise Servo would crash due to double
+    /// borrows to return it.
+    fn style_source<'a>(&'a self,
```
existing_style_for_restyle_damage?

bholley commented on components/style/animation.rs:
```diff
@@ -507,7 +507,8 @@ where Damage: TRestyleDamage {
                 if let Some(damage) = damage {
-                    *damage = *damage | Damage::compute(Some(style), &new_style);
+                    *damage =
+                        *damage | Damage::compute_for_layout(Some(style), &new_style);
```
Let's return the new style and let the caller handle it (or, if that doesn't work, let the caller provide a callback that gets passed the new style), and then remove the distinction between compute_for_layout and compute_for_cascade.

bholley commented on components/style/gecko_glue.rs:
```diff
@@ -47,6 +46,15 @@ impl<GeckoType, ServoType> ArcHelpers<GeckoType, ServoType> {
         unsafe { transmute(owned) }
     }
 
+    pub fn borrow<F, Output>(borrowed: &Arc<ServoType>, cb: F) -> Output
```
with_gecko?

bholley commented on ports/geckolib/wrapper.rs:
```diff
@@ -277,6 +313,21 @@ impl<'ln> TNode for GeckoNode<'ln> {
+
+        unsafe {
+            let context_ptr = Gecko_GetStyleContext(self.node);
+            mem::transmute(context_ptr)
```
unsafe { context_ptr.as_ref() } ?

bholley commented:
@bors-servo delegate+

r=me with comments.

bors-servo commented:
:v: @emilio can now approve this pull request

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 769cdc5 with merge 9fc5d76...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2253), [arm64](http://build.servo.org/builders/arm64/builds/2257), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2304), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2430), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2336), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2318), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2394), [windows-dev](http://build.servo.org/builders/windows-dev/builds/202)

emilio committed:
stylo: Fix and ensure alignment of restyle hints. Test it on Travis.

Turns out eRestyle_LaterSiblings was not really matching, sigh...

emilio commented:
@bholley: I'll need you to review the last commit and the regen.py changes in the previous one.

bholley commented on ports/geckolib/sanity_checks.rs:
```diff
@@ -0,0 +1,42 @@
+    use gecko_bindings::structs::nsRestyleHint;
+
+    check_enum_value_non_static!(nsRestyleHint::eRestyle_Self, RESTYLE_SELF.bits());
+    // XXX This for servo actually means something like an hipotetical
```
"a hypothetical"

bholley commented on ports/geckolib/sanity_checks.rs:
```diff
@@ -0,0 +1,42 @@
+    // XXX This for servo actually means something like an hipotetical
+    // Restyle_AllDescendants (but without running selector matching on the
+    // element). ServoRestyleManager interprets it like that, but in practice we
+    // should align the behavior with Gecko adding a new restyle hint, maybe?
```
Reference the bug we have on file for doing that?

bholley commented:
lgtm

emilio commented:
@bors-servo: r=bholley

bors-servo commented:
:pushpin: Commit 444e8f8 has been approved by `bholley`

<!-- @bors-servo r=bholley 444e8f85c8f5d0c1815f7227ec63c6b5a91f946a -->

bors-servo commented:
:hourglass: Testing commit 444e8f8 with merge cbf71a2...

bholley commented:
@bors-servo p=2

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2281), [arm64](http://build.servo.org/builders/arm64/builds/2289), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2336), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2463), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2368), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2350), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2425), [windows-dev](http://build.servo.org/builders/windows-dev/builds/234)
